infra/src/main/kotlin/fi/fta/geoviite/infra/geometry/TrackMeterHeights.kt

```diff
@@ -0,0 +1,282 @@
+    alignmentEnd: AddressPoint,
+): List<List<TrackMeterHeightTick>> {
+    val geometryAlignmentBoundariesInKmRange =
+        filterGeometryAlignmentBoundariesWithinKmRange(
+            alignment,
+            geocodingContext,
+            referencePointIndices,
+            geometryAlignmentBoundaryPoints,
+        )
+    val alignmentBoundaryTicksByKm =
+        locateAlignmentBoundaryAddresses(geometryAlignmentBoundariesInKmRange, alignment, geocodingContext)
+
+    return referencePointIndices.map { referencePointIndex ->
+        val referencePoint = geocodingContext.referencePoints[referencePointIndex]
+        val kmNumber = referencePoint.kmNumber
+        getTicksToSendForKm(
+            tickLength,
+            alignmentBoundaryTicksByKm[kmNumber] ?: listOf(),
+            referencePoint,
+            getKmLengthAtReferencePointIndex(referencePointIndex, geocodingContext),
+            kmNumber,
+            alignmentStart,
+            alignmentEnd,
+        )
+    }
+}
+
+// pure optimization: Ignore geometry alignment boundaries not within the km range we're sending
+private fun filterGeometryAlignmentBoundariesWithinKmRange(
+    alignment: IAlignment,
+    geocodingContext: GeocodingContext,
+    referencePointIndices: IntRange,
+    geometryAlignmentBoundaries: List<GeometryAlignmentBoundaryPoint>,
+): List<GeometryAlignmentBoundaryPoint> {
+    val startAddress = TrackMeter(geocodingContext.referencePoints[referencePointIndices.first].kmNumber, 0)
+    val endAddress =
+        if (referencePointIndices.last == geocodingContext.referencePoints.lastIndex) null
+        else TrackMeter(geocodingContext.referencePoints[referencePointIndices.last + 1].kmNumber, 0)
+    val rangePoints = geocodingContext.getTrackLocations(alignment, listOfNotNull(startAddress, endAddress))
+    // either or both addresses could fail to geocode onto the alignment
+    val rangeStart = rangePoints[0]?.point?.m
+    val rangeEnd = rangePoints.getOrNull(1)?.point?.m
+
+    return geometryAlignmentBoundaries.filter { boundary ->
+        (rangeStart == null || boundary.distanceOnAlignment >= rangeStart) &&
+            (rangeEnd == null || boundary.distanceOnAlignment <= rangeEnd)
+    }
+}
+
+private fun locateAlignmentBoundaryAddresses(
+    alignmentBoundaryAddresses: List<GeometryAlignmentBoundaryPoint>,
+    alignment: IAlignment,
+    geocodingContext: GeocodingContext,
+): Map<KmNumber, List<TrackMeterHeightTick>> =
+    alignmentBoundaryAddresses
+        .mapNotNull { boundary ->
+            alignment.getPointAtM(boundary.distanceOnAlignment)?.let { point ->
+                geocodingContext.getAddress(point)?.let { address -> address.first to boundary.segmentIndex }
+            }
+        }
+        .groupBy(
+            { (trackMeter) -> trackMeter.kmNumber },
+            { (trackMeter, segmentIndex) -> TrackMeterHeightTick(trackMeter.meters, segmentIndex) },
+        )
+
+private fun getTicksToSendForKm(
```

Comment:
Terminillityksenä, en oikein pidä "to send" termistä tuolla välissä noin periaatteellisena. Tässä se on tavallaan ymmärrettävä kun tätä käytetään yhdessä paikassa, mutta mikään lähettäminen on käyttäjän huolta, ei funktion. Eli ehkä vain filterTicks tai filterTicksByResolution (ticklength) tms kuulostaisi itselle paremmalta.